agent/src/workload_scheduler/scheduler.rs

```diff
@@ -0,0 +1,1482 @@
+// Copyright (c) 2024 Elektrobit Automotive GmbH
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache License, Version 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0.
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg_attr(test, mockall_double::double)]
+use crate::workload_scheduler::dependency_state_validator::DependencyStateValidator;
+
+use common::{
+    objects::{DeletedWorkload, ExecutionState, WorkloadSpec, WorkloadState},
+    std_extensions::IllegalStateResult,
+    to_server_interface::{ToServerInterface, ToServerSender},
+};
+use std::collections::HashMap;
+
+#[cfg_attr(test, mockall_double::double)]
+use crate::parameter_storage::ParameterStorage;
+use crate::workload_operation::{WorkloadOperation, WorkloadOperations};
+
+#[cfg(test)]
+use mockall::automock;
+
+#[derive(Debug, Clone, PartialEq)]
+enum PendingEntry {
+    Create(WorkloadSpec),
+    Delete(DeletedWorkload),
+    UpdateCreate(WorkloadSpec, DeletedWorkload),
+    UpdateDelete(WorkloadSpec, DeletedWorkload),
+}
+
+type WorkloadOperationQueue = HashMap<String, PendingEntry>;
+
+pub struct WorkloadScheduler {
+    queue: WorkloadOperationQueue,
+    workload_state_sender: ToServerSender,
+}
+
+#[cfg_attr(test, automock)]
+impl WorkloadScheduler {
+    pub fn new(workload_state_tx: ToServerSender) -> Self {
+        WorkloadScheduler {
+            queue: WorkloadOperationQueue::new(),
+            workload_state_sender: workload_state_tx,
+        }
+    }
+
+    async fn report_pending_create_state(&self, pending_workload: &WorkloadSpec) {
```

Comment:
Please change the vertical ordering. The public methods should be at the top and functions calls should always go downward.

Reply:
Thx, Done.